cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
 		return nil, err
 	}
 	manager := &GceManager{
-		migs:      make([]*migInformation, 0),
-		service:   gceService,
-		migCache:  make(map[GceRef]*Mig),
-		zone:      zone,
-		projectId: projectId,
+		migs:        make([]*migInformation, 0),
+		gceService:  gceService,
+		migCache:    make(map[GceRef]*Mig),
+		zone:        zone,
+		projectId:   projectId,
+		clusterName: clusterName,
+		mode:        mode,
 		templates: &templateBuilder{
 			projectId: projectId,
 			zone:      zone,
 			service:   gceService,
 		},
 	}
+
+	if mode == ModeGKE {
+		gkeService, err := gke.New(client)
+		if err != nil {
+			return nil, err
+		}
+		manager.gkeService = gkeService
+		err = manager.fetchAllNodePools()
+		if err != nil {
+			glog.Errorf("Failed to fech node pools: %v", err)
+		}
+	}
+
 	go wait.Forever(func() {
 		manager.cacheMutex.Lock()
 		defer manager.cacheMutex.Unlock()
 		if err := manager.regenerateCache(); err != nil {
 			glog.Errorf("Error while regenerating Mig cache: %v", err)
 		}
 	}, time.Hour)
+
 	return manager, nil
 }
 
-// RegisterMig registers mig in Gce Manager.
-func (m *GceManager) RegisterMig(mig *Mig) {
-	m.cacheMutex.Lock()
-	defer m.cacheMutex.Unlock()
+func (m *GceManager) assertGKE() {
+	if m.mode != ModeGKE {
+		panic(fmt.Errorf("This should run only in GKE mode"))
+	}
+}
+
+// Gets all registered node pools
+func (m *GceManager) fetchAllNodePools() error {
+	m.assertGKE()
+
+	nodePoolsResponse, err := m.gkeService.Projects.Zones.Clusters.NodePools.List(m.projectId, m.zone, m.clusterName).Do()
+	if err != nil {
+		return err
+	}
+	for _, nodePool := range nodePoolsResponse.NodePools {
+		autoprovisioned := strings.Contains("name", nodeAutoprovisioningPrefix)
+		autoscaled := nodePool.Autoscaling != nil && nodePool.Autoscaling.Enabled
+		if !autoprovisioned && !autoscaled {
+			continue
+		}
+		// format is
+		// "https://www.googleapis.com/compute/v1/projects/mwielgus-proj/zones/europe-west1-b/instanceGroupManagers/gke-cluster-1-default-pool-ba78a787-grp"
+		for _, igurl := range nodePool.InstanceGroupUrls {
+			project, zone, name, err := parseGceUrl(igurl, "instanceGroupManagers")
+			if err != nil {
+				return err
+			}
+			mig := &Mig{
+				GceRef: GceRef{
+					Name:    name,
+					Zone:    zone,
+					Project: project,
+				},
+				gceManager:      m,
+				exist:           true,
+				autoprovisioned: autoprovisioned,
+			}
+			if autoscaled {
+				mig.minSize = int(nodePool.Autoscaling.MinNodeCount)
+				mig.maxSize = int(nodePool.Autoscaling.MaxNodeCount)
+			} else if autoprovisioned {
+				mig.minSize = minAutoprovisionedSize
+				mig.maxSize = maxAutoprovisionedSize
+			}
+			m.RegisterMig(mig)
+		}
+		// TODO - unregister migs
```

Comment:
When are you planning to implement this? Sounds pretty important, especially given autoprovisioning.

Reply:
Yes. Next PR.